jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CallbackCompletionHandler.java

```diff
@@ -0,0 +1,875 @@
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.EventListener;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.component.Dumpable;
+import org.eclipse.jetty.util.component.DumpableCollection;
+import org.eclipse.jetty.util.thread.Scheduler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A troubleshooting {@link Handler.Wrapper} that tracks whether various
+ * APIs are properly used by applications.</p>
+ * <p>The violation of tracked APIs are reported to a {@link Listener}
+ * instance; the default listener implementation emits warning logs.</p>
+ * <p>{@code CallbackCompletionHandler} can be linked in at any point in
+ * the {@code Handler} chain, and even be present in multiple instances,
+ * likely configured differently.</p>
+ * <p>For example, to troubleshoot wrong usages of the callback passed to method
+ * {@link #handle(Request, Response, Callback)}, a {@code CallbackCompletionHandler}
+ * should be configured as the outermost {@code Handler}.
+ * This is because the {@code handle(...)} call propagates inwards.
+ * In this way, {@code CallbackCompletionHandler} can wrap the callback passed
+ * to inner {@code Handler}s and verify that it is eventually completed.</p>
+ * <p>On the other hand, to troubleshoot custom {@code Handler} implementations
+ * that perform wrapping of {@link Response#write(boolean, ByteBuffer, Callback)},
+ * a {@code CallbackCompletionHandler} should be configured after the custom
+ * {@code Handler} implementation.
+ * This is because the {@code write(...)} call propagates outwards.
+ * In this way, {@code CallbackCompletionHandler} can wrap the {@code write(...)}
+ * call before calling outer {@code Handler}s, and verify that it is called correctly.</p>
+ *
+ * @see Listener
+ */
+@ManagedObject
+public class CallbackCompletionHandler extends Handler.Wrapper
```

Comment:
I'm not a fan of the name, as it is doing a lot more that checking callback completion. How about:
 + `RequestStateHandler`
 + `HandlingStateHandler`
 + `HandlerDebugHandler`